data_mappings.bal

```diff
@@ -0,0 +1,462 @@
+    // if (headers is ResponseMetadata) {
+    //     user.responseHeaders = headers;
+    // }
+    return user;
+}
+
+//  Maps the json response returned from the request into record type of Permission.
+// 
+//  + jsonPayload - A tuple which contains headers and json object returned from request.
+//  + return - An instance of record type Permission.
+isolated function mapJsonToPermissionType([json, ResponseMetadata?] jsonPayload) returns @tainted Permission {
+    Permission permission = {};
+    var [payload, headers] = jsonPayload;
+    permission.id = payload.id != () ? payload.id.toString() : EMPTY_STRING;
+    permission.resourceId = payload._rid != () ? payload._rid.toString() : EMPTY_STRING;
+    permission.token = payload._token != () ? payload._token.toString() : EMPTY_STRING;
+    permission.permissionMode = payload.permissionMode != () ? payload.permissionMode.toString() : EMPTY_STRING;
+    permission.resourcePath = payload.'resource != () ? payload.'resource.toString() : EMPTY_STRING;
+    // if (headers is ResponseMetadata) {
+    //     permission.responseHeaders = headers;
+    // }
+    return permission;
+}
+
+//  Maps the json response returned from the request into record type of Offer.
+// 
+//  + jsonPayload - A tuple which contains headers and json object returned from request.
+//  + return - An instance of record type Offer.
+isolated function mapJsonToOfferType([json, ResponseMetadata?] jsonPayload) returns @tainted Offer {
+    Offer offer = {};
+    var [payload, headers] = jsonPayload;
+    offer.resourceId = payload._rid != () ? payload._rid.toString() : EMPTY_STRING;
+    offer.id = payload.id != () ? payload.id.toString() : EMPTY_STRING;
+    offer.offerVersion = payload.offerVersion != () ? payload.offerVersion.toString() : EMPTY_STRING;
+    offer.offerType = payload.offerType != () ? payload.offerType.toString() : EMPTY_STRING;
+    offer.content = payload.content != () ? payload.content.toString() : EMPTY_STRING;
+    offer.resourceSelfLink = payload.'resource != () ? payload.'resource.toString() : EMPTY_STRING;
+    offer.resourceResourceId = payload.offerResourceId != () ? payload.offerResourceId.toString() : EMPTY_STRING;
+    // if (headers is ResponseMetadata) {
+    //     offer.responseHeaders = headers;
+    // }
+    return offer;
+}
+
+//  Convert json array of database information in to an array of type Database.
+// 
+//  + databases - An existing array of type Database.
+//  + sourceDatabaseArrayJsonObject - Json object which contain the array of database information.
+//  + return - An array of type Database.
+isolated function convertToDatabaseArray(@tainted Database[] databases, json[] sourceDatabaseArrayJsonObject) returns 
+        @tainted Database[] {
+    int i = databases.length();
+    foreach json jsonDatabase in sourceDatabaseArrayJsonObject {
+        databases[i] = mapJsonToDatabaseType([jsonDatabase, ()]);
+        i = i + 1;
+    }
+    return databases;
```

Comment:
Why are we passing an array of databases, modifying it, and return the same?

Reply:
Did follow the same wrong pattern as earlier. Fixed it.